src/ws/resharder.ts

```diff
@@ -1,31 +1,141 @@
+        for (const guild of payload.guilds) {
+          gateway.resharding.markNewGuildShardId(BigInt(guild.id), shardId);
+        }
+      }
+    },
+  });
 
   // Begin resharding
   gateway.maxShards = results.shards;
+  // FOR MANUAL SHARD CONTROL, OVERRIDE THIS SHARD ID!
+  gateway.lastShardId = oldGateway.lastShardId === oldGateway.maxShards ? gateway.maxShards : oldGateway.lastShardId;
+  gateway.shardsRecommended = results.shards;
+  gateway.sessionStartLimitTotal = results.sessionStartLimit.total;
+  gateway.sessionStartLimitRemaining = results.sessionStartLimit.remaining;
+  gateway.sessionStartLimitResetAfter = results.sessionStartLimit.resetAfter;
+  gateway.maxConcurrency = results.sessionStartLimit.maxConcurrency;
   // If more than 100K servers, begin switching to 16x sharding
   if (gateway.maxShards && gateway.useOptimalLargeBotSharding) {
+    gateway.debug("[Resharding] Using optimal large bot sharding solution.");
     gateway.maxShards = Math.ceil(
       gateway.maxShards /
         (results.sessionStartLimit.maxConcurrency === 1 ? 16 : results.sessionStartLimit.maxConcurrency),
     );
   }
 
   gateway.spawnShards(gateway, gateway.firstShardId);
+
+  return new Promise((resolve) => {
+    // TIMER TO KEEP CHECKING WHEN ALL SHARDS HAVE RESHARDED
+    const timer = setInterval(async () => {
+      const pending = await gateway.resharding.isPending(gateway, oldGateway);
+      // STILL PENDING ON SOME SHARDS TO BE CREATED
+      if (pending) return;
+
+      // ENABLE EVENTS ON NEW SHARDS AND IGNORE EVENTS ON OLD
+      const oldHandler = oldGateway.handleDiscordPayload;
+      gateway.handleDiscordPayload = oldHandler;
+      oldGateway.handleDiscordPayload = function (_, data, shardId) {
+        // ALLOW EXCEPTION FOR CHUNKING TO PREVENT REQUESTS FREEZING
+        if (data.t !== "GUILD_MEMBERS_CHUNK") return;
+        oldHandler(_, data, shardId);
+      };
+
+      // STOP TIMER
+      clearInterval(timer);
+      await gateway.resharding.editGuildShardIds();
+      await gateway.resharding.closeOldShards(oldGateway);
+      gateway.debug("[Resharding] Complete.");
+      resolve(gateway);
+    }, 30000);
+  }) as Promise<GatewayManager>;
+}
+
+/** Handler that by default will check all new shards are online in the new gateway. The handler can be overriden if you have multiple servers to communicate through redis pubsub or whatever you prefer. */
+export async function resharderIsPending(
+  gateway: GatewayManager,
+  oldGateway: GatewayManager,
+) {
+  for (let i = gateway.firstShardId; i < gateway.maxShards; i++) {
+    const shard = gateway.shards.get(i);
+    if (!shard?.ready) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/** Handler that by default closes all shards in the old gateway. Can be overriden if you have multiple servers and you want to communicate through redis pubsub or whatever you prefer. */
+export async function resharderCloseOldShards(oldGateway: GatewayManager) {
```

Comment:
This will still cause a potential bug when users wait for a response from the gateway. Thoughts?

Reply:
If the gateway gets closed, then the response from discord will never come.